test/test_image.py

```diff
@@ -0,0 +1,473 @@
+    inspect_process = run(
+        split("docker inspect -f='{{.Id}}' test"),
+        capture_output=True,
+        text=True,
+        check=True,
+    )
+    return inspect_process.stdout.strip()
+
+
+def test_init(test_image_id):
+    """
+    Tests that the __init__ function on the Image class is correctly
+    receiving and remembering the ID of a docker image.
+    """
+    id = test_image_id
+    print("ID: " + id)
+    img = Image("test")
+
+    assert img is not None
+    assert img.id == id
+
+    img_2 = Image(id)
+    assert img_2.id == id
+
+    run(split("docker image remove test"))
+
+
+def test_init_CalledProcessError():
+    """
+    Tests that the __init__ function on the Image class raises a
+    CalledProcessError when given a malformed name."""
+    img = None
+    with pytest.raises(CalledProcessError):
+        img = Image("malformed_image_name_or_id")
+    assert img is None
+
+
+def test_build_from_dockerfile():
+    """
+    Tests that the build method constructs and returns an Imagemage when
+    given a Dockerfile.
+    """
+    img = Image.build(tag="test")
+    inspect_process = run(
+        split("docker inspect -f='{{.Id}}' test"),
+        text=True,
+        capture_output=True)
+    id = inspect_process.stdout.strip()
+
+    assert img is not None
+    assert img.id == id
+    run(split("docker image remove test"))
+
+
+def test_build_from_dockerfile_output_to_file():
+    """
+    Tests that the build method writes to a file when configured to
+    do so.
+    """
+    file = open("testfile.txt", "w")
+    img = Image.build(
+        tag="test",
+        dockerfile="./Dockerfile",
+        stdout=file,
+        stderr=file
+    )
+    inspect_process = run(
+        split("docker inspect -f='{{.Id}}' test"),
+        text=True,
+        capture_output=True)
+    id = inspect_process.stdout.strip()
+    file.close()
+
+    assert img is not None
+    assert img.id == id
+    assert os.stat("testfile.txt").st_size > 0
+
+    os.remove("testfile.txt")
+    run(split("docker image remove test"))
+
+
+def test_build_from_dockerfile_dockerfile_in_different_location():
+    """
+    Tests that the build method can build an image from a dockerfile in a
+    different location than the context root directory.
+    """
+    img = Image.build(
+        tag="test", dockerfile="dockerfiles/alpine_functional.dockerfile"
+    )
+    inspect_process = run(
+        split("docker inspect -f='{{.Id}}' test"),
+        text=True,
+        capture_output=True)
+    id = inspect_process.stdout.strip()
+
+    assert img is not None
+    assert img.id == id
+    run(split("docker image remove test"))
+
+
+def test_build_from_dockerfile_context_in_different_location():
+    """
+    Tests that the build method can build when the context is set to a
+    different directory.
+    """
+    img = Image.build(
+        context="./dockerfiles",
+        tag="test",
+        dockerfile="./dockerfiles/alpine_functional.dockerfile"
```

Comment:
`./` prefix should be unnecessary

```suggestion
        context="dockerfiles",
        tag="test",
        dockerfile="dockerfiles/alpine_functional.dockerfile"
```

Reply:
Does this work if you don't run pytest from within the "test" directory? 

It might be more robust to do something like this:

```python
from pathlib import Path

testdir = Path(__file__).parent
img = Image.build(
    context=testdir / "dockerfiles",
    dockerfile=testdir / "dockerfiles/alpine_functional.dockerfile",
    ...
)
```

(Note that the `dockerfile` argument can be specified as a path relative to the `context` directory, so it could just be shortened to `dockerfile="alpine_functional.dockerfile"` instead.)

Reply:
```
(Note that the dockerfile argument can be specified as a path relative to the context directory, so it could just be shortened to dockerfile="alpine_functional.dockerfile" instead.)
```
I made an attempt at this and the docker image fails to build with the following error:
```
failed to solve with frontend dockerfile.v0: failed to read dockerfile: open /var/lib/docker/tmp/buildkit-mount3558832305/alpine_functional.dockerfile: no such file or directory
Error: No such image: test
```

Reply:
Hmm weird, you're right. I can't get it to work either.

But the `docker build` documentation says this:

>By default the docker build command will look for a Dockerfile at the root of the build context. The -f, --file, option lets you specify the path to an alternative file to use instead. This is useful in cases where the same set of files are used for multiple builds. The path must be to a file within the build context. **If a relative path is specified then it is interpreted as relative to the root of the context.**

https://docs.docker.com/engine/reference/commandline/build/#text-files

Not sure what I'm missing there.